### PR TITLE
Wire in chain processing logic

### PR DESCRIPTION
**Pull request by wemeetagain**
This puts the state transition logic together into a chain module.

- eth2genesis handling is mostly complete
  - the chain module listens to the eth1 module for the eth2genesis event, then creates the genesis state/block and stores in the db
- Block processing is wired in, with minor adjustments to the underlying logic
- Fork choice is stubbed out, not currently using our lmd ghost
- needs tests for
  - genesis event
  - processing block
  - applying fork choice rule

### CODE REVIEW

**Comment by GregTheGreek** [comment]
IM SO EXCITED GREAT WORK

**Comment by GregTheGreek** [comment]
CC: @decanus THIS IS PROGRESS

File: src/chain/index.ts [hunk 1]
```diff
@@ -16,18 +30,97 @@ export class BeaconChain extends EventEmitter {
     this.eth1 = eth1;
   }
   public async start() {
```
Comment (GregTheGreek) [inline src/chain/index.ts:32]: Add space and jsdoc :)

File: src/chain/index.ts [hunk 2]
```diff
@@ -16,18 +30,97 @@ export class BeaconChain extends EventEmitter {
...
+      // listen for eth1 Eth2Genesis event
+      this.eth1.once('eth2genesis', this.processEth2Genesis.bind(this));
     }
   }
   public async stop() {}
 
-  public initializeBeaconChain(genesisTime: uint64, deposits: Deposit[], eth1Data: Eth1Data): BeaconState {
-    // eslint-disable-next-line @typescript-eslint/no-object-literal-type-assertion
-    return {} as BeaconState;
+  public async processEth2Genesis(genesisTime: uint64, genesisDeposits: Deposit[], genesisEth1Data: Eth1Data) {
```
Comment (GregTheGreek) [inline src/chain/index.ts:43]: jsdoc please :)

**Review by GregTheGreek — CHANGES_REQUESTED** [review]
First pass through main thing is make sure to get those JSDoc comments in.

I'll do a logic pass through shortly

**Comment by GregTheGreek** [comment]
Just went through all the logic one more time, everything seems very solid imo and would be good for a PR depending on how much more logic you want to add in

**Comment by wemeetagain** [comment]
Yeah, I think we should merge this and add further functionality in later PRs. I think it will be easier to debug the state transition stuff once we move our slots/epochs/validator indices from bignums to numbers.

**Comment by wemeetagain** [comment]
This should be ready for re-review.
The failing tests are from rebasing on master, looks like linter errors from the recent validator client work.

**Review by GregTheGreek — APPROVED** [review]

**Pull request merged**